services/MerchStoreService.ts

```diff
@@ -107,47 +107,69 @@ export default class MerchStoreService {
+   * options, and an item with variants enabled cannot have multiple option types.
+   */
+  private static verifyItemHasValidOptions(item: MerchItem | MerchandiseItemModel) {
+    if (!item.hasVariantsEnabled && item.options.length > 1) {
+      throw new UserError('Merch items with variants disabled cannot have multiple options');
+    }
+    if (item.hasVariantsEnabled && MerchStoreService.hasMultipleOptionTypes(item.options)) {
+      throw new UserError('Merch items cannot have multiple option types');
+    }
+  }
+
+  private static hasMultipleOptionTypes(options: MerchItemOption[]) {
+    const optionTypes = new Set(options.map((option) => option.metadata.type));
+    return optionTypes.size > 1;
+  }
+
   public async editItem(uuid: Uuid, itemEdit: MerchItemEdit): Promise<MerchandiseItemModel> {
     return this.transactions.readWrite(async (txn) => {
       const merchItemRepository = Repositories.merchStoreItem(txn);
       const item = await merchItemRepository.findByUuid(uuid);
       if (!item) throw new NotFoundError();
       const { options, collection: updatedCollection, ...changes } = itemEdit;
 
-      const merchItemOptionRepository = Repositories.merchStoreItemOption(txn);
-      const updatedOptions = await Promise.all(options.map(async (optionUpdate) => {
-        const option = await merchItemOptionRepository.findByUuid(optionUpdate.uuid);
-        if (!option) throw new NotFoundError('Item option not found');
-        // 'quantity' is incremented instead of directly set to avoid concurrency issues with orders
-        // e.g. there's 10 of an item and someone adds 5 to stock while someone else orders 1
-        // so the merch store admin sets quantity to 15 but the true quantity is 14
-        if (optionUpdate.quantity) optionUpdate.quantity += option.quantity;
-        return merchItemOptionRepository.upsertMerchItemOption(option, optionUpdate);
-      }));
+      if (options) {
+        const updatedOptionsByUuid = new Map(options.map((option) => [option.uuid, option]));
+        item.options.map((option) => {
+          if (!updatedOptionsByUuid.has(option.uuid)) return;
+          const updatedOption = updatedOptionsByUuid.get(option.uuid);
+          // 'quantity' is incremented instead of directly set to avoid concurrency issues with orders
+          // e.g. there's 10 of an item and someone adds 5 to stock while someone else orders 1
+          // so the merch store admin sets quantity to 15 but the true quantity is 14
+          if (updatedOption.addQuantity) updatedOption.addQuantity += option.quantity;
+          return MerchandiseItemOptionModel.merge(option, updatedOption);
+        });
+      }
 
-      const updatedOptionsIds = new Set(updatedOptions.map((option) => option.uuid));
-      item.options = [...updatedOptions, ...item.options.filter((option) => !updatedOptionsIds.has(option.uuid))];
+      MerchStoreService.verifyItemHasValidOptions(MerchandiseItemModel.merge(item, changes));
```

Comment:
Nit: can we move `verifyItemHasValidOptions` to below this method?

Reply:
I think it's being used in `createItem` as well, so that's why it's above. I think it's fine to leave it up there since otherwise, the reader has to scroll a bit to find the function when reading it in `createItem`